exec/java-exec/src/main/java/org/apache/drill/exec/record/RecordBatchSizer.java

```diff
@@ -536,6 +556,11 @@ public ColumnSize getColumn(String name) {
    */
   private int netRowWidth;
   private int netRowWidthCap50;
+
+  /**
+   * actual row size if input is not empty. Otherwise, standard size.
+   */
+  private int rowAllocSize;
```

Comment:
I wonder if this all would be clearer if we handed it at size estimation time. If the row count is 0, set up everything using the standard sizes. (Note: the whole reason this class exists is that the standard sizes turned out to be *very* poor estimators of actual size.)

So, if we have no data, guess the same size as `AllocationHelper`, else use real sizes.

And, again the question: under what situation do we want to use the sizer if we don't actually have any data? For the first batch, if no data, just throw away the empty batch and don't size it. Turn around and get another until we receive a non-empty batch.

If we've already received at least one non-empty batch, then we receive an empty batch, we should just retain the estimates from the non-empty batch since they will be much better than just making up numbers.

Reply:
Unless I'm missing something, we can't move forward on a join if one side is empty: we won't know if we have the rows we need. Consider a merge join (simplest). The left gets some data, but the right is empty. We can't proceed unless the right hit EOF. Otherwise, we don't know if we have a match or not for the first left row.

We need to read another right batch and keep going until we either hit EOF (no matching rows) or get some data.

Once we have some data, we can go row-by-row to see if we have a left-only, right-only, or matching set of rows. If we get to EOF on either side, we know that their are no matches for the other side.

What we do in the no-match case depends on whether we are doing LEFT OUTER, RIGHT OUTER or an INNER join.

The point is, we can't make progress until we get that non-empty right batch (in this example). So, no reason to allocate space based on an empty batch (unless the entire input is empty) because we'll need to find a non-empty (or EOF) batch anyway.

Reply:
actually, this is a problem only for lateral join. In lateral join, right side will work on one row at a time from left side. If right side produces zero rows because of empty array or some other reason, for left outer join, we still have to finish cross join for that row from left side by having nulls for right side columns. Then, we go to next row on left side, continuing to work on filling the output batch till it is full.  What that means is we have to allocate vectors based on that first batch on right side, which can be empty.

Reply:
I see. In this case, however, arrays (repeated values) will be empty. If we have 10 such rows, there is no reason to have 50 "inner" values. Also, for VarChar, no values will be stored; all columns will be null. (If we are handling non-null columns, then the non-null VarChar will be an empty string.)

So, we probably need a bit of a special case: prepare data for a run of null rows (with arrays and VarChar of length 0) vs. take our best guess with no knowledge at all about lengths (which may be non-empty.)

Probably not a huge issue if you only need to handle a single row. But, creating a batch with only one row will cause all kinds of performance issues downstream. (I found that out the hard way when a but in sort produced a series of one-row batches...)

Reply:
@paul-rogers yes, this PR is trying to do the second thing i.e. make best guess with no knowledge at all about lengths. Also, right side can produce no rows for reasons other than empty array i.e. for ex. we might be filtering everything out after unnest.  
Please let me know what you think about the approach.

Reply:
@ppadma, not much more to add. If the code requires you do estimates based on no information, we won't get very good estimates. But, if we know we have 0 rows, then that itself is a good estimate of the size we'll need.

If there is a way to improve the estimate, I'm guessing you'll find it as work proceeds and you seem more examples and test cases.